src/mito2/src/memtable/merge_tree/data.rs

```diff
@@ -396,12 +448,110 @@ pub enum DataPart {
 }
 
+pub struct DataPartIter {
+    inner: ParquetRecordBatchReader,
+    current_offset: usize,
+    current_pk_index: Option<PkIndex>,
+    current_batch: Option<RecordBatch>,
+}
+
+impl DataPartIter {
+    pub fn new(data: &[u8], batch_size: Option<usize>) -> Result<Self> {
+        let mut builder = ParquetRecordBatchReaderBuilder::try_new(Bytes::copy_from_slice(data))
+            .context(error::ReadDataPartSnafu)?;
+        if let Some(batch_size) = batch_size {
+            builder = builder.with_batch_size(batch_size);
+        }
+        let mut reader = builder.build().context(error::ReadDataPartSnafu)?;
+        let batch = reader
+            .next()
+            .transpose()
+            .context(error::ComputeArrowSnafu)?;
+        let pk_index = batch.as_ref().map(|b| pk_index_array(b).value(0));
+        Ok(Self {
+            inner: reader,
+            current_pk_index: pk_index,
+            current_offset: 0,
+            current_batch: batch,
+        })
+    }
+
+    /// Searches next primary key along with it's offset range inside record batch.
+    fn search_next_pk_range(&self) -> Option<(PkIndex, Range<usize>)> {
+        self.current_batch.as_ref().and_then(|b| {
+            // safety: PK_INDEX_COLUMN_NAME must present in record batch yielded by data part.
+            let pk_array = pk_index_array(b);
+            search_next_pk_range_(pk_array, self.current_offset)
+        })
+    }
+}
+
+impl Iterator for DataPartIter {
+    type Item = Result<DataBatch>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some((next_pk, range)) = self.search_next_pk_range() {
+            self.current_pk_index = Some(next_pk);
+            self.current_offset = range.end;
+            return Some(Ok(DataBatch {
+                pk_index: next_pk,
+                rb: self.current_batch.as_ref().unwrap().clone(), // safety: current batch won't be none.
+                range,
+            }));
+        } else if let Some(res) = self.inner.next() {
+            let batch = match res {
+                Ok(b) => b,
+                Err(e) => {
+                    return Some(Err(e).context(error::ComputeArrowSnafu));
+                }
+            };
+            self.current_batch = Some(batch);
+            self.current_offset = 0;
+            self.next()
+        } else {
+            return None;
+        }
+    }
+}
+
 impl DataPart {
-    pub fn read(&self, _pk_weights: &[u16]) -> Result<DataBatch> {
-        todo!()
+    /// Iterates frozen data parts and yields record batches.
+    /// Returned record batches are ga
+    pub fn iter(&self, _pk_weights: &[u16]) -> Result<DataPartIter> {
+        match self {
+            DataPart::Parquet(data_bytes) => DataPartIter::new(data_bytes, None),
+        }
     }
 }
 
+/// Searches for next pk index and it's offset range in a sorted `UInt16Array`.
+fn search_next_pk_range_(array: &UInt16Array, start: usize) -> Option<(PkIndex, Range<usize>)> {
```

Comment:
No need to add suffix to the function, it is different from `DataPartIter::search_next_pk_range()`.
```suggestion
fn search_next_pk_range(array: &UInt16Array, start: usize) -> Option<(PkIndex, Range<usize>)> {
```